receiver/pulsarreceiver/pulsar_receiver.go

```diff
@@ -0,0 +1,328 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulsarreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/pulsarreceiver"
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/zap"
+)
+
+var errUnrecognizedEncoding = errors.New("unrecognized encoding")
+
+const alreadyClosedError = "AlreadyClosedError"
+
+type pulsarTracesConsumer struct {
+	id              config.ComponentID
+	tracesConsumer  consumer.Traces
+	topic           string
+	client          pulsar.Client
+	cancel          context.CancelFunc
+	consumer        pulsar.Consumer
+	unmarshaler     TracesUnmarshaler
+	settings        component.ReceiverCreateSettings
+	consumerOptions pulsar.ConsumerOptions
+}
+
+func newTracesReceiver(config Config, set component.ReceiverCreateSettings, unmarshalers map[string]TracesUnmarshaler, nextConsumer consumer.Traces) (*pulsarTracesConsumer, error) {
+	unmarshaler := unmarshalers[config.Encoding]
+	if nil == unmarshaler {
+		return nil, errUnrecognizedEncoding
+	}
+
+	options := config.clientOptions()
+	client, err := pulsar.NewClient(options)
+	if err != nil {
+		return nil, err
+	}
+
+	consumerOptions, err := config.consumerOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	return &pulsarTracesConsumer{
+		id:              config.ID(),
+		tracesConsumer:  nextConsumer,
+		topic:           config.Topic,
+		unmarshaler:     unmarshaler,
+		settings:        set,
+		client:          client,
+		consumerOptions: consumerOptions,
+	}, nil
+}
+
+func (c *pulsarTracesConsumer) Start(context.Context, component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancel = cancel
+
+	_consumer, err := c.client.Subscribe(c.consumerOptions)
+	if err == nil {
+		c.consumer = _consumer
+		go func() {
+			e := consumerTracesLoop(ctx, c)
+			if e != nil {
+				c.settings.Logger.Error("Consume tracesConsumer failed", zap.Error(e))
+			}
+		}()
+	}
+
+	return err
+}
+
+func consumerTracesLoop(ctx context.Context, c *pulsarTracesConsumer) error {
+	unmarshaler := c.unmarshaler
+	traceConsumer := c.tracesConsumer
+
+	for {
+		message, err := c.consumer.Receive(ctx)
+		if err != nil {
+			if strings.Contains(err.Error(), alreadyClosedError) {
+				return err
+			}
+			if errors.Is(err, context.Canceled) {
+				c.settings.Logger.Info("exiting consume traces loop")
+				return err
+			}
+			time.Sleep(time.Second)
+			c.settings.Logger.Error("failed to receive traces message from Pulsar", zap.Error(err))
+			continue
+		}
+
+		traces, err := unmarshaler.Unmarshal(message.Payload())
+		if err != nil {
+			c.settings.Logger.Error("failed to unmarshaler traces message", zap.Error(err))
```

Comment:
The message should be acknowledged otherwise the receiver will stuck trying to fetch the same broken record, right?